pkg/controller/cluster/cluster_controller.go

```diff
@@ -151,20 +213,91 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
+			return reconcile.Result{}, errors.WithStack(err)
 		}
 
 		return reconcile.Result{}, nil
 	}
 
-	klog.Infof("reconciling cluster object %v triggers idempotent reconcile.", name)
+	klog.Infof("Reconciling cluster object %v triggers idempotent reconcile.", name)
 	if err := r.actuator.Reconcile(cluster); err != nil {
 		if requeueErr, ok := errors.Cause(err).(controllerError.HasRequeueAfterError); ok {
 			klog.Infof("Actuator returned requeue-after error: %v", requeueErr)
 			return reconcile.Result{Requeue: true, RequeueAfter: requeueErr.GetRequeueAfter()}, nil
 		}
 		klog.Errorf("Error reconciling cluster object %v; %v", name, err)
-		return reconcile.Result{}, err
+		return reconcile.Result{}, errors.WithStack(err)
 	}
 	return reconcile.Result{}, nil
 }
+
+// listChildren returns a list of Deployments, Sets, and Machines than have an ownerref to the given cluster
+func (r *ReconcileCluster) listChildren(ctx context.Context, cluster *clusterv1.Cluster) ([]runtime.Object, error) {
+	var children []runtime.Object
+
+	ns := cluster.GetNamespace()
+	opts := metav1.ListOptions{
+		LabelSelector: labels.FormatLabels(
+			map[string]string{clusterv1.MachineClusterLabelName: cluster.GetName()},
+		),
+	}
+
+	dfunc := func(ctx context.Context, m metav1.ListOptions) (runtime.Object, error) {
+		return r.clusterClient.MachineDeployments(ns).List(m)
+	}
+	sfunc := func(ctx context.Context, m metav1.ListOptions) (runtime.Object, error) {
+		return r.clusterClient.MachineSets(ns).List(m)
+	}
+	mfunc := func(ctx context.Context, m metav1.ListOptions) (runtime.Object, error) {
+		return r.clusterClient.Machines(ns).List(m)
+	}
+
+	eachFunc := func(o runtime.Object) error {
+		acc, err := meta.Accessor(o)
+		if err != nil {
+			return err
+		}
+
+		if pointsTo(acc.GetOwnerReferences(), &cluster.ObjectMeta) {
+			children = append(children, o)
+		}
+
+		return nil
+	}
+
+	deployments, err := pager.New(dfunc).List(ctx, opts)
+	if err != nil {
+		return []runtime.Object{}, errors.Wrapf(err, "failed to list MachineDeployments in %s/%s", ns, cluster.Name)
+	}
+	if err := meta.EachListItem(deployments, eachFunc); err != nil {
+		return []runtime.Object{}, errors.Wrapf(err, "couldn't iterate MachinesDeployments for cluster %s/%s", ns, cluster.Name)
+	}
+
+	sets, err := pager.New(sfunc).List(ctx, opts)
+	if err != nil {
+		return []runtime.Object{}, errors.Wrapf(err, "failed to list MachineSets in %s/%s", ns, cluster.Name)
+	}
+	if err := meta.EachListItem(sets, eachFunc); err != nil {
+		return []runtime.Object{}, errors.Wrapf(err, "couldn't iterate MachineSets for cluster %s/%s", ns, cluster.Name)
+	}
+
+	machines, err := pager.New(mfunc).List(ctx, opts)
+	if err != nil {
+		return []runtime.Object{}, errors.Wrapf(err, "failed to list Machines in %s/%s", ns, cluster.Name)
+	}
+	if err := meta.EachListItem(machines, eachFunc); err != nil {
+		return []runtime.Object{}, errors.Wrapf(err, "couldn't iterate Machines for cluster %s/%s", ns, cluster.Name)
```

Comment:
Can we return nil instead of an empty slice in when we have errors?

Reply:
huh, TIL